File: examples/13-connectors/main.go
// examples/13-connectors/main.go demonstrates the three connector types in gopptx.
//
// Shows straight, elbow, and curved connectors placed on separate slides
// alongside the shapes they connect.
//
// Run with: go run ./examples/13-connectors/main.go
package main

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/djinn-soul/gopptx/pkg/pptx"
	"github.com/djinn-soul/gopptx/pkg/pptx/styling"
	log "github.com/djinn-soul/gopptx/pkg/stdlog"
)

const (
	outputDir  = "examples/output"
	outputFile = "13_connectors.pptx"
)

func main() {
	if err := run(); err != nil {
		fmt.Fprintf(os.Stderr, "error: %v\n", err)
		os.Exit(1)
	}
}

func run() error {
	if err := os.MkdirAll(outputDir, 0o750); err != nil {
		return fmt.Errorf("create output directory: %w", err)
	}

	// --- Slide 1: overview ---
	slide1 := pptx.NewSlide("Connectors Demo").
		AddBullet("gopptx supports three connector types:").
		AddBullet("  • Straight — direct line between two points").
		AddBullet("  • Elbow   — right-angle (bent) connector").
		AddBullet("  • Curved  — smooth bezier-curve connector")

	// --- Slide 2: straight connector ---
	straight := pptx.NewStraightConnector(
		styling.Inches(1), styling.Inches(2), // startX, startY
		styling.Inches(5), styling.Inches(2), // endX, endY
	)

	leftBox := pptx.NewShape(pptx.ShapeTypeRectangle,
		styling.Inches(0.25), styling.Inches(1.75),
		styling.Inches(0.75), styling.Inches(0.5),
	).WithFill(pptx.NewShapeFill("4472C4")).WithText("A")

	rightBox := pptx.NewShape(pptx.ShapeTypeRectangle,
		styling.Inches(5), styling.Inches(1.75),
		styling.Inches(0.75), styling.Inches(0.5),
	).WithFill(pptx.NewShapeFill("4472C4")).WithText("B")

	slide2 := pptx.NewSlide("Straight Connector").
		AddShape(leftBox).
		AddShape(rightBox).
		AddConnector(straight).
		AddBullet("pptx.NewStraightConnector(startX, startY, endX, endY)")

	// --- Slide 3: elbow connector ---
	elbow := pptx.NewElbowConnector(
		styling.Inches(1), styling.Inches(2),
		styling.Inches(5), styling.Inches(4),
	)

	topBox := pptx.NewShape(pptx.ShapeTypeRectangle,
		styling.Inches(0.25), styling.Inches(1.75),
		styling.Inches(0.75), styling.Inches(0.5),
	).WithFill(pptx.NewShapeFill("ED7D31")).WithText("C")

	bottomBox := pptx.NewShape(pptx.ShapeTypeRectangle,
		styling.Inches(5), styling.Inches(3.75),
		styling.Inches(0.75), styling.Inches(0.5),
	).WithFill(pptx.NewShapeFill("ED7D31")).WithText("D")

	slide3 := pptx.NewSlide("Elbow Connector").
		AddShape(topBox).
		AddShape(bottomBox).
		AddConnector(elbow).
		AddBullet("pptx.NewElbowConnector(startX, startY, endX, endY)")

	// --- Slide 4: curved connector ---
	curved := pptx.NewCurvedConnector(
		styling.Inches(1), styling.Inches(3),
		styling.Inches(5), styling.Inches(5),
	)

	topLeft := pptx.NewShape(pptx.ShapeTypeEllipse,
		styling.Inches(0.25), styling.Inches(2.75),
		styling.Inches(0.75), styling.Inches(0.5),
	).WithFill(pptx.NewShapeFill("A9D18E")).WithText("E")

	bottomRight := pptx.NewShape(pptx.ShapeTypeEllipse,
		styling.Inches(5), styling.Inches(4.75),
		styling.Inches(0.75), styling.Inches(0.5),
	).WithFill(pptx.NewShapeFill("A9D18E")).WithText("F")

	slide4 := pptx.NewSlide("Curved Connector").
		AddShape(topLeft).
		AddShape(bottomRight).
		AddConnector(curved).
		AddBullet("pptx.NewCurvedConnector(startX, startY, endX, endY)")

	slides := []pptx.SlideContent{slide1, slide2, slide3, slide4}

	outputPath := filepath.Join(outputDir, outputFile)
	if err := pptx.WriteFile(outputPath, "Connectors Demo", slides); err != nil {
		return fmt.Errorf("write presentation: %w", err)
	}

	log.Printf("Generated %s\n", outputPath)
	return nil
}

File: examples/13-connectors/main.py
"""Demonstrate the three connector types: straight, elbow, and curved.

This example demonstrates:
- Straight connector between two points
- Elbow (right-angle) connector
- Curved (bezier) connector
- Shapes placed alongside connectors to show connection points
"""

from __future__ import annotations

from pathlib import Path

from gopptx import Presentation
from gopptx.presentation.slides import SlideLayoutType
from gopptx.schemas import Inches


def _add_overview_slide(prs: Presentation) -> None:
    """Add an overview slide listing the connector types."""
    prs.add_bullet_slide(
        "Connectors Demo",
        [
            "gopptx supports three connector types:",
            "  Straight — direct line between two points",
            "  Elbow   — right-angle (bent) connector",
            "  Curved  — smooth bezier-curve connector",
        ],
    )


def _add_straight_connector_slide(prs: Presentation) -> None:
    """Add a slide demonstrating a straight connector."""
    prs.add_slide("Straight Connector", layout=SlideLayoutType.BLANK)
    idx = prs.slide_count - 1

    prs.add_shape(
        idx,
        "RECTANGLE",
        bounds=(Inches(0.25), Inches(1.75), Inches(0.75), Inches(0.5)),
        text="A",
        properties={"fill": {"solid": "4472C4"}},
    )
    prs.add_shape(
        idx,
        "RECTANGLE",
        bounds=(Inches(5), Inches(1.75), Inches(0.75), Inches(0.5)),
        text="B",
        properties={"fill": {"solid": "4472C4"}},
    )
    prs.add_connector(idx, "STRAIGHT", Inches(1), Inches(2), Inches(5), Inches(2))
    prs.add_textbox(
        idx,
        Inches(0.5),
        Inches(3),
        Inches(6),
        Inches(0.5),
        text="add_connector(idx, 'STRAIGHT', x1, y1, x2, y2)",
    )


def _add_elbow_connector_slide(prs: Presentation) -> None:
    """Add a slide demonstrating an elbow connector."""
    prs.add_slide("Elbow Connector", layout=SlideLayoutType.BLANK)
    idx = prs.slide_count - 1

    prs.add_shape(
        idx,
        "RECTANGLE",
        bounds=(Inches(0.25), Inches(1.75), Inches(0.75), Inches(0.5)),
        text="C",
        properties={"fill": {"solid": "ED7D31"}},
    )
    prs.add_shape(
        idx,
        "RECTANGLE",
        bounds=(Inches(5), Inches(3.75), Inches(0.75), Inches(0.5)),
        text="D",
        properties={"fill": {"solid": "ED7D31"}},
    )
    prs.add_connector(idx, "ELBOW", Inches(1), Inches(2), Inches(5), Inches(4))
    prs.add_textbox(
        idx,
        Inches(0.5),
        Inches(5),
        Inches(6),
        Inches(0.5),
        text="add_connector(idx, 'ELBOW', x1, y1, x2, y2)",
    )


def _add_curved_connector_slide(prs: Presentation) -> None:
    """Add a slide demonstrating a curved connector."""
    prs.add_slide("Curved Connector", layout=SlideLayoutType.BLANK)
    idx = prs.slide_count - 1

    prs.add_shape(
        idx,
        "ELLIPSE",
        bounds=(Inches(0.25), Inches(2.75), Inches(0.75), Inches(0.5)),
        text="E",
        properties={"fill": {"solid": "A9D18E"}},
    )
    prs.add_shape(
        idx,
        "ELLIPSE",
        bounds=(Inches(5), Inches(4.75), Inches(0.75), Inches(0.5)),
        text="F",
        properties={"fill": {"solid": "A9D18E"}},
    )
    prs.add_connector(idx, "CURVED", Inches(1), Inches(3), Inches(5), Inches(5))
    prs.add_textbox(
        idx,
        Inches(0.5),
        Inches(6),
        Inches(6),
        Inches(0.5),
        text="add_connector(idx, 'CURVED', x1, y1, x2, y2)",
    )


def main() -> None:
    """Create a presentation demonstrating all three connector types."""
    output_dir = Path("examples/output")
    output_dir.mkdir(exist_ok=True)

    with Presentation.new("Connectors Demo") as prs:
        _add_overview_slide(prs)
        _add_straight_connector_slide(prs)
        _add_elbow_connector_slide(prs)
        _add_curved_connector_slide(prs)

        output_path = output_dir / "13-connectors.pptx"
        prs.save(str(output_path))
        print(f"Saved: {output_path}")

    print("\n=== SUMMARY ===")
    print("Created 4 slides demonstrating connector types:")
    print("  - Overview")
    print("  - Straight connector")
    print("  - Elbow (right-angle) connector")
    print("  - Curved connector")


if __name__ == "__main__":
    main()
